ledger/src/blockstore_metrics.rs

```diff
@@ -516,7 +512,45 @@ pub(crate) fn report_rocksdb_write_perf(metric_header: &'static str) {
 /// A struct that holds the current status of RocksDB perf sampling.
 pub struct PerfSamplingStatus {
     // The number of RocksDB operations since the last perf sample.
-    pub(crate) op_count: AtomicUsize,
+    op_count: AtomicUsize,
+    // The timestamp of the latest operation with perf stats collection.
+    last_sample_time_ms: AtomicU64,
+}
+
+impl PerfSamplingStatus {
+    fn should_sample(&self, sample_interval: usize) -> bool {
+        if sample_interval == 0 {
+            return false;
+        }
+
+        if self.op_count.fetch_add(1, Ordering::Relaxed) < sample_interval {
+            return false;
+        }
+        // Ideally, fetch_sub(*sample_interval) should be used to keep it
+        // super precise.  However, since we do not use Mutex to protect the
+        // above check and the below operation, we simply reset it to 0.
+        self.op_count.store(0, Ordering::Relaxed);
+
+        // Rate-limiting the perf sampling based on the minimum time interval.
+        let current_time_ms = timestamp();
+        let old_time_ms = self.last_sample_time_ms.load(Ordering::Relaxed);
+        if old_time_ms + (PERF_SAMPLING_MIN_DURATION.as_millis() as u64) < current_time_ms {
+            return false;
+        }
+
+        // If the `last_sample_time_ms` has a different value than `old_time_ms`,
+        // it means some other thread has performed the sampling and updated
+        // the last sample time.  In this case, the current thread will skip
+        // the current sample.
+        self.last_sample_time_ms
+            .compare_exchange_weak(
```

Comment:
Thanks, @behzadnouri.  It looks a lot better now!

So now the check is still the same using one atomic load, but the update is done via `compare_and_exchange` so that it only returns true if the value is unchanged when we performed compare_and_exchange.